addons/mail/models/res_users.py

```diff
@@ -268,42 +268,45 @@ def _init_messaging(self):
     @api.model
     def systray_get_activities(self):
         activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
-        activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
-        for activity in activities:
-            record = self.env[activity.res_model].browse(activity.res_id)
-            activities_by_record_by_model_name[activity.res_model][record] += activity
-        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_record_by_model_name.keys()})
-        user_activities = {}
-        for model_name, activities_by_record in activities_by_record_by_model_name.items():
-            domain = [("id", "in", list({r.id for r in activities_by_record.keys()}))]
-            allowed_records = self.env[model_name].search(domain)
-            if not allowed_records:
-                continue
-            module = self.env[model_name]._original_module
-            icon = module and modules.module.get_module_icon(module)
-            model = self.env["ir.model"]._get(model_name).with_prefetch(model_ids)
-            user_activities[model_name] = {
-                "id": model.id,
-                "name": model.name,
-                "model": model_name,
-                "type": "activity",
-                "icon": icon,
-                "total_count": 0,
-                "today_count": 0,
-                "overdue_count": 0,
-                "planned_count": 0,
-                "actions": [
-                    {
-                        "icon": "fa-clock-o",
-                        "name": "Summary",
-                    }
-                ],
-            }
-            for record, activities in activities_by_record.items():
-                if record not in allowed_records:
+        activities_by_rec_id_by_module_by_model_name = {}
+        group_custom_info = {}
+        models = list(dict.fromkeys(activities.mapped('res_model')))
+        for model in models:
+            activities_by_rec_id_by_module_by_model_name[model], group_custom_info[model] = self.env[model]._sort_activities_by_record_by_module(
+                activities.filtered_domain([('res_model', '=', model)]))
+        model_ids = list({self.env["ir.model"]._get(name).id for name in activities_by_rec_id_by_module_by_model_name})
+        user_activities = []
+        for model_name, activities_by_rec_id_by_module in activities_by_rec_id_by_module_by_model_name.items():
+            domain = [("id", "in", list(set.union(*[{id for id in activities_by_rec_id} for activities_by_rec_id in activities_by_rec_id_by_module.values()])))]
+            allowed_records_ids = self.env[model_name].search(domain).ids
+            for module, activities_by_rec_id in activities_by_rec_id_by_module.items():
+                if not allowed_records_ids:
                     continue
```

Comment:
We could even do this:
```suggestion
            allowed_records_ids = self.env[model_name].search(domain).ids
            if not allowed_records_ids:
                continue
            for module, activities_by_rec_id in activities_by_rec_id_by_module.items():
```
no? 🧐

Reply:
> We could even do this:
> 
> no? monocle_face

As discussed, my last change was not good because some empty groups could be created. So I've changed that part taking your comment into account, thx :+1: